### PR TITLE
Improve algorithm to generate a filename from an image download

### PR DESCRIPTION
**Pull request by CDRussell**
Task/Issue URL: https://app.asana.com/0/0/1187737416677993/f
Tech Design URL: 
CC: 

**Description**:
When trying to download images, our app currently tries to guess what a good filename for the download will be, and often fails to produce something nice; it'll often end in a .*bin file type as its current implementation is somewhat naive in turning a URL into a filename.

This PR improves that situation by walking back the URL path, trying each to make a filename, and then returning the best filename produced. Worst case, this should produce the same result as the old algorithm, but best case it will pick out a much better filename.

**Steps to test this PR**:
1. Visit https://vignette3.wikia.nocookie.net/guns/images/b/b1/FHJ84.jpg/revision/latest?cb=20160824141549 and long press to download the image; verify the filename produced is `FHJ84.jpg` and not `latest.bin` as it was before
1. Try other image downloads, and verify they produce a sensible enough filename

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: app/src/main/java/com/duckduckgo/app/browser/downloader/FilenameExtractor.kt [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        if (latestGuess.contains(".")) {
+            guesses.bestGuess = latestGuess
+        }
+
+        if (pathSegments.size < 2) return TriedAllOptions
+
+        return NotGoodEnough
+    }
+
+    private fun guessFromUrl(url: String, contentDisposition: String?, mimeType: String?): String {
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/browser/downloader/FilenameExtractor.kt:55]: That made me think this method was going to use only `url: String` to guess the fileName, but I think we will also use also the other params. Wanted to check if my assumption is true

File: app/src/main/java/com/duckduckgo/app/browser/downloader/FileDownloader.kt [hunk 2]
```diff
@@ -27,7 +27,8 @@ import javax.inject.Inject
 
 class FileDownloader @Inject constructor(
     private val dataUriDownloader: DataUriDownloader,
-    private val networkFileDownloader: NetworkFileDownloader
+    private val networkFileDownloader: NetworkFileDownloader,
+    private val filenameExtractor: FilenameExtractor
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/browser/downloader/FileDownloader.kt:31]: unused param `filenameExtractor`

File: app/src/main/java/com/duckduckgo/app/browser/downloader/FileDownloader.kt [hunk 3]
```diff
@@ -57,8 +58,8 @@ class FileDownloader @Inject constructor(
     }
 }
 
-fun FileDownloader.PendingFileDownload.guessFileName(): String {
-    val guessedFileName = URLUtil.guessFileName(url, contentDisposition, mimeType)
+fun FileDownloader.PendingFileDownload.guessFileName(filenameExtractor: FilenameExtractor): String {
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/browser/downloader/FileDownloader.kt:61]: No strong opinion, and not really important, just talking out loud.

It feels a bit weird that `PendingFileDownload` has a method `guessFileName` that receives a `FileNameExtractor` to guess the file name. I would expect the other way around: `FileNameExtractor.guessFileNameFrom(pendingFileDownload)` as a extension function that receives a `PendingFileDownload`. Inside that method we map `PendingFileDownload` to `(url, contentDisposition, mimeType)`

File: app/src/main/java/com/duckduckgo/app/browser/downloader/FilenameExtractor.kt (hunk 1 above)
Comment (CDRussell) [reply app/src/main/java/com/duckduckgo/app/browser/downloader/FilenameExtractor.kt]: Good point; changed to `guessFilename()` which is hopefully clearer

File: app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        assertEquals("realFilename.bin", extracted)
+    }
+
+    @Test
+    fun whenUrlContainsNoFileNameButLotsOfPathsSegmentsThenFirstSegmentNameIsUsed() {
+        val url = "https://example.com/foo/bar/files"
+        val mimeType: String? = null
+        val contentDisposition: String? = null
+        val extracted = testee.extract(url, contentDisposition, mimeType)
+        assertEquals("foo", extracted)
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt]: Fixed to return `.bin` if all no filetype available on the best guess

File: app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.browser.downloader
+
+import org.junit.Assert.assertEquals
+import org.junit.Test
+
+class UriUtilsFilenameExtractorTest {
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt:22]: 1. `contentDisposition` is preferred. that is unchanged from current behavior; we still fall back on `URLUtil.guessFileName` which decides on the precedence between them. I added a unit test in here.
1. I see `mimetype` being preferred, but again, deferring that logic to `URLUtil.guessFileName`. so for this scenario, the behavior is unchanged from the current behavior.

File: app/src/main/java/com/duckduckgo/app/browser/downloader/FileDownloader.kt (hunk 3 above)
Comment (CDRussell) [reply app/src/main/java/com/duckduckgo/app/browser/downloader/FileDownloader.kt]: Agree, this is weird. I've removed that extension function on `pendingFileDownload` and put the logic all in `FileNameExtractor`

**Pull request merged**